patches/net/minecraft/world/effect/MobEffectInstance.java.patch

```diff
@@ -86,16 +86,23 @@
                              .apply(p_324063_, MobEffectInstance.Details::create)
                  )
          );
-+        // TODO: https://github.com/neoforged/NeoForge/issues/986 - Fix this to have vanilla-compatible sync.
-         public static final StreamCodec<ByteBuf, MobEffectInstance.Details> STREAM_CODEC = StreamCodec.recursive(
-             p_329990_ -> StreamCodec.composite(
+-        public static final StreamCodec<ByteBuf, MobEffectInstance.Details> STREAM_CODEC = StreamCodec.recursive(
+-            p_329990_ -> StreamCodec.composite(
++        public static final StreamCodec<RegistryFriendlyByteBuf, MobEffectInstance.Details> STREAM_CODEC = StreamCodec.recursive(
++            p_329990_ -> net.neoforged.neoforge.network.codec.NeoForgeStreamCodecs.composite(
                      ByteBufCodecs.VAR_INT,
-@@ -427,14 +_,23 @@
+                     MobEffectInstance.Details::amplifier,
+                     ByteBufCodecs.VAR_INT,
+@@ -427,14 +_,20 @@
                      MobEffectInstance.Details::showIcon,
                      p_329990_.apply(ByteBufCodecs::optional),
                      MobEffectInstance.Details::hiddenEffect,
-+                    // net.neoforged.neoforge.common.EffectCure.STREAM_CODEC.<java.util.Set<net.neoforged.neoforge.common.EffectCure>>apply(ByteBufCodecs.collection(java.util.HashSet::new)).apply(ByteBufCodecs::optional),
-+                    // MobEffectInstance.Details::cures,
++                    // Neo: Add additional serialization logic for custom EffectCure(s)
++                    net.neoforged.neoforge.network.codec.NeoForgeStreamCodecs.connectionAware(
++                            ByteBufCodecs.optional(net.neoforged.neoforge.common.EffectCure.STREAM_CODEC.apply(ByteBufCodecs.collection(java.util.HashSet::new))),
++                            net.neoforged.neoforge.network.codec.NeoForgeStreamCodecs.uncheckedUnit(Optional.empty())
```

Comment:
How does this behave if a vanilla server connects to a Neo client? Will the client automatically delegate the missing effect cures to the default defined in Java?

Reply:
Upon receipt, the client will reach the `uncheckedUnit` codec, and return `Optional.empty()` without doing any further accesses to the buffer data

Reply:
I understand that part, but when I try to access the set of cures in this client only mod, will I get an empty set or will I get the default set?

Reply:
> I understand that part, but when I try to access the set of cures in this client only mod, will I get an empty set or will I get the default set?

You'll get the empty set, that's the current design flaw of the EffectCure system, which I have already made a issue #1060 proposing for changing that.

Reply:
My understanding was that issue was for NBT parsing. Network parsing is a part of that issue that may be reasonable to fix in this PR.

Reply:
We may want to at least return `DEFAULT_CURES` here instead of nothing.  I don't think we have the context in this object to return potion specific ones (honey for poison).  We _might_ be able to do that one level above (in the `MobEffectInstance` codec rather than `Details`) with some post-processing magic.

Reply:
I believe cures that are identical with defaults should never be serialized for good, either in persistent data or networking.
See #1064.

Reply:
> We may want to at least return `DEFAULT_CURES` here instead of nothing. I don't think we have the context in this object to return potion specific ones (honey for poison). We _might_ be able to do that one level above (in the `MobEffectInstance` codec rather than `Details`) with some post-processing magic.

The alternative is we just interpret an empty optional as default in the getter, and call some default method on the effect instance there

Reply:
Resolving this conversation as of #1064, discussion for default cures could be moved to there.